cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go

```diff
@@ -102,70 +137,164 @@ func (m *asgCache) Unregister(asg *Asg) bool {
 	return changed
 }
 
-func (m *asgCache) get() []*asgInformation {
+func (m *asgCache) buildAsgFromSpec(spec string) (*asg, error) {
+	s, err := dynamic.SpecFromString(spec, scaleToZeroSupported)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse node group spec: %v", err)
+	}
+	asg := &asg{
+		AwsRef:  AwsRef{Name: s.Name},
+		minSize: s.MinSize,
+		maxSize: s.MaxSize,
+	}
+	return asg, nil
+}
+
+// Get returns the currently registered ASGs
+func (m *asgCache) Get() []*asg {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
 	return m.registeredAsgs
 }
 
 // FindForInstance returns AsgConfig of the given Instance
-func (m *asgCache) FindForInstance(instance *AwsRef) (*Asg, error) {
-	// TODO(negz): Prevent this calling describe ASGs too often.
+func (m *asgCache) FindForInstance(instance AwsInstanceRef) (*asg, error) {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	if m.notInRegisteredAsg[*instance] {
-		// We already know we don't own this instance. Return early and avoid
-		// additional calls to describe ASGs.
-		return nil, nil
+	if asg, found := m.instanceToAsg[instance]; found {
+		return asg, nil
 	}
 
-	if config, found := m.instanceToAsg[*instance]; found {
-		return config, nil
-	}
-	if err := m.regenerate(); err != nil {
-		return nil, fmt.Errorf("Error while looking for ASG for instance %+v, error: %v", *instance, err)
-	}
-	if config, found := m.instanceToAsg[*instance]; found {
-		return config, nil
+	return nil, nil
+}
+
+// InstancesByAsg returns the nodes of an ASG
+func (m *asgCache) InstancesByAsg(ref AwsRef) ([]AwsInstanceRef, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if instances, found := m.asgToInstances[ref]; found {
+		return instances, nil
 	}
 
-	m.notInRegisteredAsg[*instance] = true
-	return nil, nil
+	return nil, fmt.Errorf("Error while looking for instances of ASG: %s", ref)
 }
 
-func (m *asgCache) invalidateUnownedInstanceCache() {
-	glog.V(4).Info("Invalidating unowned instance cache")
-	m.notInRegisteredAsg = make(map[AwsRef]bool)
+// Fetch automatically discovered ASGs. These ASGs should be unregistered if
+// they no longer exist in AWS.
+func (m *asgCache) fetchAutoAsgNames() ([]string, error) {
+	groupNames := make([]string, 0)
+
+	for _, spec := range m.asgAutoDiscoverySpecs {
+		names, err := m.service.getAutoscalingGroupNamesByTags(spec.Tags)
+		if err != nil {
+			return nil, fmt.Errorf("cannot autodiscover ASGs: %s", err)
+		}
+
+		groupNames = append(groupNames, names...)
+	}
+
+	return groupNames, nil
 }
 
+// regenerate the cached view of explicitly configured and auto-discovered ASGs
```

Comment:
This method is getting pretty huge at 70+ lines. I wonder if it's worth breaking out a little. 🤔